dice_ml/explainer_interfaces/dice_genetic.py

```diff
@@ -94,94 +94,75 @@ def do_loss_initializations(self, yloss_type, diversity_loss_type, feature_weigh
+            for jx, feature in enumerate(self.data_interface.feature_names):
+                if feature in features_to_vary:
+                    if feature in self.data_interface.continuous_feature_names:
+                        one_init[jx] = np.round(np.random.uniform(self.feature_range[feature][0], self.feature_range[feature][1]), precisions[jx])
                     else:
-                        temp_cfs.append(np.array(one_init))
-                ix += 1
-            self.cfs.append(temp_cfs)
+                        one_init[jx] = np.random.choice(self.feature_range[feature])
+                else:
+                    one_init[jx] = query_instance[jx]
+            if self.model.model_type == 'classifier':
+                if self.predict_fn(np.array(one_init)) != desired_class:
+                    kx -= 1
+                else:
+                    remaining_cfs[kx] = one_init
+            elif self.model.model_type == 'regressor':
+                predicted_value = self.predict_fn(np.array(one_init))
+                if not desired_range[0] <= predicted_value <= desired_range[1]:
+                    kx -= 1
+                else:
+                    remaining_cfs[kx] = one_init
+            kx += 1
+        return remaining_cfs
 
-    def do_KD_init(self, features_to_vary, query_instance, cfs):
+    def do_KD_init(self, features_to_vary, query_instance, cfs, desired_class, desired_range):
         cfs = self.label_encode(cfs)
         cfs = cfs.reset_index(drop=True)
-        ix = 0
-        done = False
+
+        self.cfs = np.zeros((self.population_size, self.data_interface.number_of_features))
         for kx in range(self.population_size):
-            temp_cfs = []
-            for _ in range(self.total_CFs):
-                if ix >= len(cfs):
-                    done = True
-                    break
-                one_init = [[]]
-                for jx, feature in enumerate(self.data_interface.feature_names):
-                    if feature not in features_to_vary:
-                        one_init[0].append(query_instance[0][jx])
-                    else:
-                        if feature in self.data_interface.continuous_feature_names:
-                            if self.feature_range[feature][0] <= cfs.iloc[ix][jx] <= self.feature_range[feature][1]:
-                                one_init[0].append(cfs.iloc[ix][jx])
+            if kx >= len(cfs):
+                break
+            one_init = np.zeros(self.data_interface.number_of_features)
+            for jx, feature in enumerate(self.data_interface.feature_names):
+                if feature not in features_to_vary:
+                    one_init[jx] = (query_instance[jx])
+                else:
+                    if feature in self.data_interface.continuous_feature_names:
+                        if self.feature_range[feature][0] <= cfs.iloc[kx][jx] <= self.feature_range[feature][1]:
+                            one_init[jx] = cfs.iloc[kx][jx]
```

Comment:
use `at` or `iat` instead of loc/iloc everywhere in the code. It is much faster. https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html _"For production code, we recommended that you take advantage of the optimized pandas data access methods exposed in this chapter."_